Core/src/Detector/VolumeStructureBuilder.cpp

```diff
@@ -0,0 +1,190 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2023 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "Acts/Detector/VolumeStructureBuilder.hpp"
+
+#include "Acts/Detector/PortalGenerators.hpp"
+#include "Acts/Geometry/ConeVolumeBounds.hpp"
+#include "Acts/Geometry/CuboidVolumeBounds.hpp"
+#include "Acts/Geometry/CutoutCylinderVolumeBounds.hpp"
+#include "Acts/Geometry/CylinderVolumeBounds.hpp"
+#include "Acts/Geometry/GenericCuboidVolumeBounds.hpp"
+#include "Acts/Geometry/TrapezoidVolumeBounds.hpp"
+#include "Acts/Utilities/Enumerate.hpp"
+
+namespace {
+
+/// This can be abandoned with C++20 to use the std::to_array method
+///
+/// @note only the first kDIM elments will obviously be filled, if the
+/// vector tends to be longer, it is truncated
+///
+/// @param vecvals the vector of bound values to be converted
+/// @return an array with the filled values
+template <std::size_t kDIM>
+std::array<Acts::ActsScalar, kDIM> to_array(
+    const std::vector<Acts::ActsScalar>& vecvals) {
+  std::array<Acts::ActsScalar, kDIM> rarray = {};
+  for (const auto [iv, v] : Acts::enumerate(vecvals)) {
+    if (iv < kDIM) {
+      rarray[iv] = v;
+    }
+  }
+  return rarray;
+}
+}  // namespace
+
+Acts::Experimental::VolumeStructureBuilder::VolumeStructureBuilder(
+    const Acts::Experimental::VolumeStructureBuilder::Config& cfg,
+    std::unique_ptr<const Acts::Logger> logger)
+    : IExternalStructureBuilder(), m_cfg(cfg), m_logger(std::move(logger)) {
+  // Sanity cross-checks
+  if (m_cfg.boundValues.empty() and not m_cfg.extent.has_value()) {
+    throw std::invalid_argument(
+        "VolumeStructureBuilder: no extent nor boundary values given");
+  }
+  // Check for the bounds type
+  if (m_cfg.boundsType > VolumeBounds::BoundsType::eTrapezoid) {
+    throw std::invalid_argument(
+        "VolumeStructureBuilder: no bounds type provided.");
+  }
+}
+
+Acts::Experimental::ExternalStructure
+Acts::Experimental::VolumeStructureBuilder::construct(
+    [[maybe_unused]] const Acts::GeometryContext& gctx) const {
+  // The volume bounds to be constructed
+  std::unique_ptr<VolumeBounds> volumeBounds = nullptr;
+
+  // A transform
+  auto transform = Transform3::Identity();
+  std::vector<ActsScalar> boundValues = m_cfg.boundValues;
+
+  // This code dispatches into the dedicated volume types
+  switch (m_cfg.boundsType) {
+    case VolumeBounds::BoundsType::eCone: {
+      ACTS_VERBOSE("Building conical volume bounds.");
+      // Cone translation - only pre-defined values
+      if (boundValues.size() < 5u) {
+        throw std::runtime_error(
+            "VolumeStructureBuilder: parameters for cone volume bounds need to "
```

Comment:
Since the error message is thrown several times in slight variations, you might consider making a lambda that returns the string to make the code slicker, something like:
`
 auto notEnoughParamsMessage =
      [](std::string boundTypeStr) {
        return "VolumeStructureBuilder: parameters for generic " + boundTypeStr +
            " volume bounds need to be fully provided, they can not be estimated from an Extent object."
      }
`

Reply:
Probably you don't even need the lambda, since this is guarded by the `if` anyway.